### PR TITLE
fix: don't crash controller on reconcile panics

### PR DESCRIPTION
**Pull request by jessesuen**
Partially resolves https://github.com/akuity/kargo/issues/656

Depends on https://github.com/akuity/kargo/pull/650

If a promotion function panics, we will fail the promotion rather than crash the controller, and the promotion will have a proper message and phase:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Promotion
metadata:
  creationTimestamp: "2023-09-06T21:11:10Z"
  generation: 1
  name: dev.01h9p5bczmwqj1z5hhf6880p8f.e9b3b60
  namespace: kargo-guestbook
  ownerReferences:
  - apiVersion: kargo.akuity.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: Stage
    name: dev
    uid: 32626a1f-c00f-468a-8e76-d1a0d83e7920
  resourceVersion: "127330362"
  uid: 5482cac2-16ed-4f5f-ad55-2d3a4e5c8db7
spec:
  freight: e9b3b603e17ea2e291535b605a58c2e62f1949be
  stage: dev
status:
  error: 'runtime error: invalid memory address or nil pointer dereference'
  phase: Failed
```

### CODE REVIEW

**Comment by jessesuen** [comment]
I should rebase this ontop of https://github.com/akuity/kargo/pull/650.

**Comment by krancour** [comment]
Do we want this change long term? If we do it here, idk why we wouldn't be doing it all over the place, and obviously we won't because this isn't customary in Go.

So I'm suggesting that maybe we leave a comment here that notes this can be removed when Bookkeeper is fixed.

Otherwise LGTM

**Comment by jessesuen** [comment]
> Do we want this change long term? If we do it here, idk why we wouldn't be doing it all over the place, and obviously we won't because this isn't customary in Go.

We do. One single bad resource should not cause system-wide unavailability. The risk of not doing this means complete downtime / denial of service.  In fact this is a common enough scenario that api-machinery controller runtime [has a helper for this](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/runtime#HandleCrash) (but I don't want to use it here since we need to fail the promotion rather than requeue and retry reconciliation later).

**Comment by krancour** [comment]
Fair point... is it then worthwhile to apply something similar to the other two reconcilers? (Stages and Applications)

**Comment by jessesuen** [comment]
> Fair point... is it then worthwhile to apply something similar to the other two reconcilers? (Stages and Applications)

Yea we should. I wasn't sure if those were impacted if they were built using kubebuilder (I know the promo reconciler was custom built). I'll go check those other reconcilers and improve those if they crash the controller.

**Comment by krancour** [comment]
> I wasn't sure if those were impacted if they were built using kubebuilder

Yea... kubebuilder feels overly-opinionated in regard to overall project structure, so we just use controller runtime directly.

**Comment by jessesuen** [comment]
OK, I improved it so all reconcilers will recover from panic. Looks like controller-runtime has made this even easier than it used to be, by introducing a `RecoverPanic: true` flag:

```
	return controller.Options{
		RecoverPanic: true,
	}
```

**Comment by krancour** [comment]
> Looks like controller-runtime has made this even easier than it used to be, by introducing a RecoverPanic: true flag

Beautiful!

File: internal/controller/promotions/promotions.go [hunk 1]
```diff
@@ -326,16 +328,26 @@ func (r *reconciler) serializedSync(
...
+						promoCtx,
+						promo.Spec.Stage,
+						promo.Namespace,
+						promo.Spec.Freight,
+					); err != nil {
+						phase = kargoapi.PromotionPhaseErrored
+						phaseError = err.Error()
+						logger.Errorf("error executing Promotion: %s", err)
+					}
+				}()
```
Comment (krancour) [inline internal/controller/promotions/promotions.go:350]: Are these changes still required in light of the recover panic option?

Comment (jessesuen) [reply internal/controller/promotions/promotions.go:350]: Yes I intentionally left this in, because this has the added benefit of patching the promotion so that promo enters a Error state. Without this, the Promos that cause the panic, will continue to reconcile in a loop until the crash is dealt with. Unlike Stages/Applications (which do have to reconcile forever), at least with Promos we can stop the bleeding.

Comment (krancour) [reply internal/controller/promotions/promotions.go:350]: That makes sense!

**Review by krancour — APPROVED** [review]
